accesslog/middleware_gin.go

```diff
@@ -0,0 +1,147 @@
+// Copyright 2020 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+package accesslog
+
+import (
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/gin-gonic/gin"
+	"github.com/mendersoftware/go-lib-micro/log"
+)
+
+type LogParameters struct {
+	// Context contains the entire request context as exposed by
+	// gin.Context.
+	Context *gin.Context
+
+	// Path holds the path as set before the request handler is executed.
+	Path string
+
+	// StartTime is the time when the request was received by the
+	// middleware, which can be used to compute latency.
+	StartTime time.Time
+}
+
+type LogHook func(parms LogParameters)
+
+func defaultLogHook(params LogParameters) {
+	latency := time.Since(params.StartTime)
+	c := params.Context
+	l := log.FromContext(c.Request.Context())
+	code := c.Writer.Status()
+	// Add status and response time to log context
+	logCtx := log.Ctx{
```

Comment:
The logger should be compatible with what we already have in workflows:
https://github.com/mendersoftware/workflows/blob/master/api/http/logger.go#L45

The reason is that our logstash rules are based on that set of fields.

Reply:
ahh, ok. But won't `clientip` simply point to the gateway anyway? Or do we send `X-Forwarded-(IP|Host|For)` headers?

Reply:
And do we have to provide a `msg`? It doesn't add any additional information to the log entry except repeating existing fields.

Reply:
no, we do not need the msg, and it is automatically removed by logstash. It is there because of back-compatibility when we switched to the new logstash parsing. Regarding clientIP, I confirm it is not the real clientIP, and there is no need at this stage to manage the `X-Forwarded-X` headers.